File: 278.First Bad Version.cpp
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int min_version = 1;
        int max_version = n;
        
        if (isBadVersion(1)) {
            return 1;
        }

        while (min_version+1 != max_version) {
            int bound = (max_version - min_version) /2 + min_version;
            if(!isBadVersion(bound))
                min_version = bound;
            else
                max_version = bound;
        }
        
        return max_version;
    }
};

File: LeetCode/278.First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        if isBadVersion(1): return 1
        if not isBadVersion(n): return -1
        start, end = 1, n
        while start < end:
            mid = (start + end) / 2
            mid_result = isBadVersion(mid)
            next_result = isBadVersion(mid + 1)
            if not mid_result and next_result:
                return mid + 1
            elif mid_result:
                end = mid
            else:
                start = mid + 1
        return -1
